### PR TITLE
Infer binding sources for ApiController parameters

### LINKED ISSUE #6847: For types annotated with ApiControllerAttribute, consider making complex type parameters FromBody by default
As part of producing, more opinionated APIs via `ApiControllerAttribute` we can assume any complex type parameter in your action without binding information is interpreted as `FromBody`. This will reduce the amount of typing \ clutter in user code:

```C#
[ApiController]
public class OrdersController
{
   // ...
   [HttpPost]
   public IActionResult Create(Book book); // Assumes FromBody

   [HttpPut]
   public IActionResult Update([FromQuery] Book book); // Opt out
}
```

We could add a switch to `ApiControllerOptions` to disable this behavior globally.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #6847

### CODE REVIEW

**Comment by pranavkm** [comment]
I'll add more tests \ functionals. Getting it out there to verify design.

File: src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelMetadata.cs [hunk 1]
```diff
@@ -325,7 +325,7 @@ public virtual ModelMetadata ContainerMetadata
         public Type ElementType { get; private set; }
 
         /// <summary>
-        /// Gets a value indicating whether <see cref="ModelType"/> is a simple type.
+        /// Gets a value indicating whether <see cref="ModelType"/> is a complex type.
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelMetadata.cs]: Oops 😆

File: src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelMetadata.cs [hunk 2]
```diff
@@ -325,7 +325,7 @@ public virtual ModelMetadata ContainerMetadata
...
         /// </summary>
         /// <remarks>
         /// A simple type is defined as a <see cref="Type"/> which has a
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelMetadata.cs]: This should probably be updated too

File: src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelMetadataProvider.cs [hunk 3]
```diff
@@ -31,7 +30,7 @@ public abstract class ModelMetadataProvider : IModelMetadataProvider
         /// Supplies metadata describing a parameter.
         /// </summary>
         /// <param name="parameter">The <see cref="ParameterInfo"/>.</param>
-        /// <returns>A <see cref="ModelMetadata"/> instance describing properties of the <see cref="ActionDescriptor"/>.</returns>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelMetadataProvider.cs]: lol x2

File: src/Microsoft.AspNetCore.Mvc.Core/ApiBehaviorOptions.cs [hunk 4]
```diff
@@ -28,10 +28,21 @@ public class ApiBehaviorOptions
         }
 
         /// <summary>
-        /// Disables the filter that returns an <see cref="BadRequestObjectResult"/> when
-        /// <see cref="ActionContext.ModelState"/> is invalid. 
+        /// Gets or sets a value that determines if the filter that returns an <see cref="BadRequestObjectResult"/> when
+        /// <see cref="ActionContext.ModelState"/> is invalid. Defaults to <c>true</c>.
         /// <seealso cref="InvalidModelStateResponseFactory"/>.
         /// </summary>
         public bool EnableModelStateInvalidFilter { get; set; } = true;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ApiBehaviorOptions.cs]: Sorry I didn't catch this in the past, this should be called `SuppressModelStateInvalidFilter` and default to false. We want boolean options to default to false, it's really surprising when something has a non-default value.

File: src/Microsoft.AspNetCore.Mvc.Core/ApiBehaviorOptions.cs [hunk 5]
```diff
@@ -28,10 +28,21 @@ public class ApiBehaviorOptions
...
+        /// <summary>
+        /// Gets or sets a value that determines if model binding sources are inferred for parameters that do not explicitly
+        /// specify one.
+        /// <para>
+        /// Parameters that appear as route values, are assumed to be bound from the path (<see cref="BindingSource.Path"/>).
+        /// Parameters that are complex (<see cref="ModelMetadata.IsComplexType"/>) are assumed to be bound from the body (<see cref="BindingSource.Body"/>).
+        /// All other parameters are assumed to be bound from the query.
+        /// </para>
+        /// </summary>
+        public bool InferBindingSourcesForParameters { get; set; } = true;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ApiBehaviorOptions.cs]: Same thing here, invert the positrons

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        private readonly ModelStateInvalidFilter _modelStateInvalidFilter;
+        private readonly ILogger _logger;
+
+        public ApiBehaviorApplicationModelProvider(
+            IOptions<ApiBehaviorOptions> apiBehaviorOptions,
+            IModelMetadataProvider modelMetadataProvider,
+            ILoggerFactory loggerFactory)
+        {
+            _apiBehaviorOptions = apiBehaviorOptions.Value;
+            _modelMetadataProvider = modelMetadataProvider as ModelMetadataProvider;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs]: What if this fails?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        {
+        }
+
+        public void OnProvidersExecuting(ApplicationModelProviderContext context)
+        {
+            foreach (var controllerModel in context.Result.Controllers)
+            {
+                var isApiController = controllerModel.Attributes.OfType<IApiBehaviorMetadata>().Any();
+                var controllerHasSelectorModel = controllerModel.Selectors.Any(s => s.AttributeRouteModel != null);
+
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs]: extra line

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+                    if (!isApiController && !actionModel.Attributes.OfType<IApiBehaviorMetadata>().Any())
+                    {
+                        continue;
+                    }
+
+                    if (!controllerHasSelectorModel && !actionModel.Selectors.Any(s => s.AttributeRouteModel != null))
+                    {
+                        // Require attribute routing with controllers annotated with ApiControllerAttribute
+                        throw new InvalidOperationException(Resources.FormatApiController_AttributeRouteRequired(nameof(ApiControllerAttribute)));
+                    }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs]: Can we separate all of these 'logical operations' into private methods? I feel like this list of steps is going to grow and it's already at like 3.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+                    if (_apiBehaviorOptions.EnableModelStateInvalidFilter)
+                    {
+                        Debug.Assert(_apiBehaviorOptions.InvalidModelStateResponseFactory != null);
+                        actionModel.Filters.Add(_modelStateInvalidFilter);
+                    }
+
+                    if (_modelMetadataProvider != null && _apiBehaviorOptions.InferBindingSourcesForParameters)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs]: What about bound properties? IMO we should do the same thing

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+                return BindingSource.Path;
+            }
+            else
+            {
+                var parameterMetadata = _modelMetadataProvider.GetMetadataForParameter(parameter.ParameterInfo);
+                if (parameterMetadata != null)
+                {
+                    var bindingSource = parameterMetadata.IsComplexType ?
+                        BindingSource.Body :
+                        BindingSource.Query;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs]: What is the experience like if you end up with multiple `[FromBody]` parameters? Can you easily figure out what's wrong?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+                    return false;
+                }
+
+                // Ensure at least one route exists.
+                parameterExistsInSomeRoute = true;
+            }
+
+            return parameterExistsInSomeRoute;
+
+            IEnumerable<string> GetAllRouteTemplates()
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs]: Is this code a dupe of something else? Should we consolidate this or provide extension methods?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcCoreLoggerExtensions.cs [hunk 12]
```diff
@@ -289,6 +293,10 @@ static MvcCoreLoggerExtensions()
                 1,
                 "The request has model state errors, returning an error response.");
 
+            _inferringParameterSource = LoggerMessage.Define<MethodInfo, string, string>(
+                LogLevel.Debug,
+                1,
+                "Inferring binding source for '{ActionName}.{ParameterName} as {BindingSource}.");
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcCoreLoggerExtensions.cs]: `Inferred`

Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcCoreLoggerExtensions.cs]: `'{ActionName}.{ParameterName` - this is missing a closing quote. Also this isn't a good syntax - no one expects `Foo.Bar` to mean parameter

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs (hunk 9 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs]: Isn't that problematic? We'd be looking at properties without any BindingInfo and assigning them binding sources. Plus we'd have to know about other things that are ModelBinding-ish (like `TempData`) and make we're not attempting to model bind them.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs (hunk 10 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs]: Would bailing at that point with Log.Info \ Log.Warn make sense?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs (hunk 11 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs]: It's a clone of https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs#L147. I'll see if I can refactor it to be re-used.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs [hunk 13]
```diff
@@ -137,96 +137,38 @@ public static IList<ControllerActionDescriptor> Build(ApplicationModel applicati
             ControllerModel controller,
             ActionModel action)
         {
-            var controllerAttributeRoutes = controller.Selectors
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: I'm not entirely sure what this check was meant for. Essentially, this ensures at least one controller selector has an attribute route associated. If it is, the method combined it with all controller selectors. I would think we only want to merge it with controller selectors that have an associated attribute route. Doing so causes one of our unit tests to fail.

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: Right now, I'm just ignoring this check - https://github.com/aspnet/Mvc/pull/6930/files#diff-f9ecc4fe4438037c1ffb1d25eac827f3R30

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+            for (var i = 0; i < inferredBindingSources.Length; i++)
+            {
+                var parameter = actionModel.Parameters[i];
+                var bindingSource = InferBindingSourceForParameter(parameter);
+                if (bindingSource == BindingSource.Body)
+                {
+                    if (foundFromBodyParameter)
+                    {
+                        // More than one parameter is inferred as FromBody. Log a warning and skip this action.
+                        _logger.UnableToInferBindingSource(actionModel);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs]: Can you log an issue to decide if this should be an error instead? (throwing)

I don't want to block you on this.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs (hunk 10 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs]: Let's roll this into a separate issue

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs (hunk 13 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: If you're still unsure about this it's going to require some more digging. A concrete example to discuss would really help.

I would be super super super careful making any changes to this code.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs (hunk 10 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/ApiBehaviorApplicationModelProvider.cs]: https://github.com/aspnet/Mvc/issues/6963

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs (hunk 13 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: Fair enough. I'll leave the thing in and file a separate issue for this.

**Pull request closed without merging**